Modules/pyexpat.c

```diff
@@ -1631,101 +1627,152 @@ static int init_handler_descrs(void)
+    PyObject *codes_dict = NULL, *rev_codes_dict = NULL;
+
+    pyexpat_state *state = pyexpat_get_state(mod);
+    state->xml_parse_type = (PyTypeObject *)PyType_FromModuleAndSpec(
+        mod, &_xml_parse_type_spec, NULL);
+    if (state->xml_parse_type == NULL) {
+        goto error;
+    }
+
+    if (init_handler_descrs(state) < 0) {
+        goto error;
+    }
 
     /* Add some symbolic constants to the module */
-    if (ErrorObject == NULL) {
-        ErrorObject = PyErr_NewException("xml.parsers.expat.ExpatError",
-                                         NULL, NULL);
-        if (ErrorObject == NULL)
-            return NULL;
+    state->error = PyErr_NewException("xml.parsers.expat.ExpatError",
+                                        NULL, NULL);
+    if (state->error == NULL) {
+        goto error;
+    }
+    
+    Py_INCREF(state->error);
+    if (PyModule_AddObject(mod, "error", state->error ) < 0) {
+        goto error;
+    }
+    Py_INCREF(state->error);
+    if (PyModule_AddObject(mod, "ExpatError", state->error ) < 0) {
+        goto error;
+    }
+    Py_INCREF(state->xml_parse_type);
+    if (PyModule_AddObject(mod, "XMLParserType",
+                           (PyObject *) state->xml_parse_type) < 0) {
+        goto error;
     }
-    Py_INCREF(ErrorObject);
-    PyModule_AddObject(m, "error", ErrorObject);
-    Py_INCREF(ErrorObject);
-    PyModule_AddObject(m, "ExpatError", ErrorObject);
-    Py_INCREF(&Xmlparsetype);
-    PyModule_AddObject(m, "XMLParserType", (PyObject *) &Xmlparsetype);
 
-    PyModule_AddStringConstant(m, "EXPAT_VERSION",
-                               XML_ExpatVersion());
+    if (PyModule_AddStringConstant(mod, "EXPAT_VERSION",
+                               XML_ExpatVersion()) < 0) {
+        goto error;
+    }
     {
         XML_Expat_Version info = XML_ExpatVersionInfo();
-        PyModule_AddObject(m, "version_info",
-                           Py_BuildValue("(iii)", info.major,
-                                         info.minor, info.micro));
+        if (PyModule_AddObject(mod, "version_info", Py_BuildValue("(iii)",
+                               info.major, info.minor, info.micro)) < 0) {
+            goto error;
+        }
     }
     /* XXX When Expat supports some way of figuring out how it was
        compiled, this should check and set native_encoding
        appropriately.
     */
-    PyModule_AddStringConstant(m, "native_encoding", "UTF-8");
+    if (PyModule_AddStringConstant(mod, "native_encoding", "UTF-8") < 0) {
+        goto error;
+    }
 
-    d = PyModule_GetDict(m);
+    PyObject *d = PyModule_GetDict(mod);
     if (d == NULL) {
-        Py_DECREF(m);
-        return NULL;
+        goto error;
+    }
+
+    PyObject *errmod_name = PyUnicode_FromString(MODULE_NAME ".errors");
+    if (errmod_name == NULL) {
+        goto error;
     }
-    errors_module = PyDict_GetItemWithError(d, errmod_name);
+
+    PyObject *errors_module = PyDict_GetItemWithError(d, errmod_name);
```

Comment:
IMO PyDict_GetItemWithError(d, errmod_name) can be removed, it's a bug and is useless.

This PR converts pyexpat to multiphase init. Each time the exec function is called, we get a fresh module instance.

I don't understand when pyexpat.__dict__ dictionary contains the "pyexpat.errors" key. Even after PyModule_AddObject(mod, "errors", errors_module) below, "pyexpat.errors" key still doesn't exist. The key is "errors".

It looks like it's an error in an old version of the code. Example in Python 3.5:

```
    sys_modules = PySys_GetObject("modules");
    d = PyModule_GetDict(m);
    errors_module = PyDict_GetItem(d, errmod_name);
    if (errors_module == NULL) {
        errors_module = PyModule_New(MODULE_NAME ".errors");
        if (errors_module != NULL) {
            PyDict_SetItem(sys_modules, errmod_name, errors_module);
            /* gives away the reference to errors_module */
            PyModule_AddObject(m, "errors", errors_module);
        }
    }
    Py_DECREF(errmod_name);
    model_module = PyDict_GetItem(d, modelmod_name);
    if (model_module == NULL) {
        model_module = PyModule_New(MODULE_NAME ".model");
        if (model_module != NULL) {
            PyDict_SetItem(sys_modules, modelmod_name, model_module);
            /* gives away the reference to model_module */
            PyModule_AddObject(m, "model", model_module);
        }
    }
    Py_DECREF(modelmod_name);
```

``pyexpat.errors`` is the correct key for sys.modules dictionary. But we should not touch sys.modules anymore in the new code. It's better to only work on a module object.

Reply:
@vstinner it looks like the intention in Python 3.5 was to do `d = PyModule_GetDict(sys_modules);`

I can't say I understand the rationale but I can remove it to simplify this ugly block.